Issue 460: Fail on getfuncargvalue(<fixture with params>)

tomviner opened the pull request:
Fix for https://github.com/pytest-dev/pytest/issues/460

Raise helpful failure message, when requesting parametrized fixture at runtime, e.g. with ``request.getfuncargvalue``.

This now fails before getting to `AttributeError: 'SubRequest' object has no attribute 'param'`.

**BACKWARD INCOMPAT**: Previously these params were simply never defined. So a fixture decorated like `@pytest.fixture(params=[0, 1, 2])` only runs once. Now a failure is raised.

RonnyPfannschmidt commented on _pytest/python.py:
```diff
@@ -1651,6 +1651,21 @@ def _getfuncargvalue(self, fixturedef):
+            if fixturedef.params is not None:
+                curdir = py.path.local()
+                frame = inspect.stack()[3]
+                _, source_path, source_lineno, _, _, _ = frame
```
source-path should be relative to the test root if possible

The-Compiler replied:
This seems kind of weird - if I'm reading the docs correctly you can do `inspect.getframeinfo(frame)` and then have a namedtuple with `filename` and `lineno` attributes.

RonnyPfannschmidt commented on _pytest/python.py:
```diff
@@ -1651,6 +1651,21 @@ def _getfuncargvalue(self, fixturedef):
             param = NOTSET
             param_index = 0
+            if fixturedef.params is not None:
+                curdir = py.path.local()
```
use  the test root here

nicoddemus commented on testing/python/fixture.py:
```diff
@@ -2723,3 +2723,75 @@ def test_1(arg1):
+            E*Failed: The requested fixture has no parameter defined for the current test.
+            E*
+            E*Requested fixture 'fix_with_param' defined in:
+            E*{}:4
```
This is not valid in `py26`, that's why CI failed.

nicoddemus commented on CHANGELOG.rst:
```diff
@@ -17,14 +17,20 @@
 
-*
+* Raise helpful failure message, when requesting parametrized fixture at runtime,
+  e.g. with ``request.getfuncargvalue``. BACKWARD INCOMPAT: Previously these params
```
As we discussed during the planning today, we would like to change that to `getfixturevalue` (keeping the older name only for backward compatibility). Would you like to do this in this PR? If not, would you please open a new issue? Thanks! :grin:

tomviner replied:
Opened https://github.com/pytest-dev/pytest/issues/1625

tomviner commented:
Added the commented changes
* used named tuple from inspect.getframeinfo
* made path relative to root dir and added test
* fixed python 2.6 .format needed {0}
* changed change log entry

The-Compiler commented:
LGTM, but I'd rather not be the one deciding on the backwards compatibility break and if it's okay :laughing: - as said, I think it's perfectly fine for 3.0.

nicoddemus commented:
>  be the one deciding on the backwards compatibility break and if it's okay

Hmm which backward compatibility break?

nicoddemus commented:
>  be the one deciding on the backwards compatibility break and if it's okay

Hmm which backward compatibility break?

tomviner commented:
@The-Compiler @RonnyPfannschmidt 

>> be the one deciding on the backwards compatibility break and if it's okay

> Hmm which backward compatibility break?

The fact that this previously silently failed to apply the params:
```python
@pytest.fixture(params=[0, 1, 2])
def fix_with_param(request):
    print "I never call request.param"

def test_foo(request):
    request.getfuncargvalue('fix_with_param')
```

And now fails with the helpful error message.
